MQMQA RelaxNG implementation

bocklund opened the pull request:
Still needs Python code to hook up read/write support.

bocklund committed:
MQMQA rng prototype

bocklund committed:
Add Chemsage file for testing

bocklund commented on xml2/MQMQA.rng:
```diff
@@ -1,32 +1,162 @@
+                        <value>MQMX</value><a:documentation>Excess Gibbs energy for a quadruplet</a:documentation>
+                    </choice>
                 </attribute>
+                <!--TODO: can the different parameter types be part of the <choice> tag?"-->
```
It might help for the parameter specific options to have a `<choice>` of `<group>` so that the optional parameter-specific tags are mutually exclusive. 

As an alternative, maybe there needs to be a `<choice>` of `<element name=“Parameter”>` tags, instead of allowing types to be mixed-and-matched with different specialty options (like MQMZ and the Exponents element, which would be wrong).

bocklund committed:
MQMQA: implement writer and schema updates

Schema adds "version" (SUBG or SUBQ) and adds two exponents so there are four total (like in the Database representation)

bocklund committed:
Add simple test module

bocklund committed:
Make the additional parameter metadata be a <choice> instead of sequential <optional>. Also interleaves parameter metadata to fix a element ordering bug

bocklund commented:
Made some minor changes to the schema and `write_xml` is working now according to the schema. I also added some code to validate the XML in the writer function and raise with the errors if invalid.

Needs implementation for the reader function for MQMQA and then a little bit of additional work to add in the CEF Kohler-Toop mixing parameters for QKTO support.

bocklund committed:
Delete stray print statements

bocklund committed:
Fix MQMQA constituent writing

bocklund committed:
Writing MQMQA with roundtrip calculation test

- Add all stoichiometric factors to XML
- Spelling correction

bocklund committed:
Add better roundtrip and roundtrip-roundtrip tests

bocklund committed:
Fix round-round-trip

bocklund committed:
Switch exponents to float

bocklund commented:
I have some writing support now.

In my `test_mqmqa_write.py` module, I have a test that does a `read_DAT` -> `write_xml` -> `read_XML` test that passes for database equality and passes a functional test of the energy for MQMQA.

I also have a `read_DAT` -> `write_xml` -> `read_XML` -> `write_xml` -> `read_XML` test that helped find a validation bug. The bug was related to Exponents.

This needs the changes from https://github.com/pycalphad/pycalphad/pull/405 to be able to read and write correctly.

bocklund commented:
- This needs an extension to handle the QKTO case still. 
- In the current formulation, users can probably modify the XML database to insert their own `<Expr>` tags and use them in the MQM parameters (`MQMG` and `MQMX`). With that, one should be able to use ESPEI to fit the energy coefficients and read-write XML files (assuming a mechanism to register the functions from this repo - probably a modified pycalphad). This functionality still needs to be tested.

bocklund committed:
Rename parser test module

bocklund committed:
Add some roundtrip tests for TDBs

richardotis commented:
> * In the current formulation, users can probably modify the XML database to insert their own `<Expr>` tags and use them in the MQM parameters (`MQMG` and `MQMX`). With that, one should be able to use ESPEI to fit the energy coefficients and read-write XML files (assuming a mechanism to register the functions from this repo - probably a modified pycalphad).

What mechanism do you mean is needed? Shouldn't any `<Expr>` tags get picked up in `dbf.symbols`?

bocklund committed:
Fix parameter_order naming bug

bocklund committed:
Add QKTO read/write to XML parser (CEF schema)

- Add choice between "Order" and "Exponents" for QKT parameters
- Add chemical groups support

bocklund committed:
Add simple qkto test while debugging

bocklund committed:
Fix NCName type for phase names to be able to parse DAT phases with parenthesis in the name.

Test with FeTiVO read/write instead of simple-qkto now that we can support writing/reading the whole thing.

bocklund committed:
Use Kaye database for QKTO test

bocklund commented:
> > * In the current formulation, users can probably modify the XML database to insert their own `<Expr>` tags and use them in the MQM parameters (`MQMG` and `MQMX`). With that, one should be able to use ESPEI to fit the energy coefficients and read-write XML files (assuming a mechanism to register the functions from this repo - probably a modified pycalphad).
> 
> What mechanism do you mean is needed? Shouldn't any `<Expr>` tags get picked up in `dbf.symbols`?

Sorry, by "register the function" I mean that a user of ESPEI that wants to test this will have to somehow run:

```python
Database.register_format("xml", read=read_xml, write=write_xml)
```

Using a modified version of pycalphad with a one line change to import the `xml_parser` module seems like a logical way to do that. 

Any symbol references to `<Expr>` tags in `param["parameter"]` expressions should just work unless I messed something up.

bocklund commented:
With the latest commits, I think I have working MQMQA and QKTO read and write now. I can round trip compare equal to a few DAT file sin the `test_xml_parser.py` module. Needs the latest changes to https://github.com/pycalphad/pycalphad/pull/405 to pull in some DAT parser bug fixes.

richardotis commented:
> Sorry, by "register the function" I mean that a user of ESPEI that wants to test this will have to somehow run:
> 
> ```python
> Database.register_format("xml", read=read_xml, write=write_xml)
> ```
> 
> Using a modified version of pycalphad with a one line change to import the `xml_parser` module seems like a logical way to do that.

Why wouldn't this work?

```python
import xml_parser
import espei
# etc
```
Or do you mean running ESPEI from the command line using its entry point script?

bocklund commented:
Correct, for the entry point. I am still running ESPEI from the command line most of the time. 

As a workaround, it should work to do something like 

`python -m xml_parser -m espei.espei_script --in mcmc.yaml`

It’s not the ideal experience, but doesn’t require any code changes.

bocklund committed:
Additional mixing code support

bocklund committed:
Temporary workaround for import from another working directory

bocklund committed:
Add require_valid parameter to write_xml to be able to write invalid XML files

bocklund commented:
@richardotis I think this one is done and ready for you to review and/or merge as you'd like. All the DAT files that I have locally that pycalphad can parse are able to round trip to/from XML with validation.

richardotis committed:
Fix typos
